File: simple-message/src/main/java/ru/starksoft/simplemessage/SimpleMessageManager.java
package ru.starksoft.simplemessage;

import android.os.Build;
import android.os.Handler;
import android.os.Looper;
import android.support.annotation.NonNull;
import android.support.annotation.UiThread;
import android.util.Log;

import static ru.starksoft.simplemessage.util.Util.checkNotNull;

@SuppressWarnings("WeakerAccess")
final class SimpleMessageManager {

	private static final String TAG = "SimpleMessageManager";

	private static final SimpleMessageManager INSTANCE = new SimpleMessageManager();

	private final Handler handler = new Handler(Looper.getMainLooper());

	private Message currentMessage;

	private SimpleMessageManager() {
	}

	@NonNull
	public static SimpleMessageManager getInstance() {
		return INSTANCE;
	}

	@UiThread
	public void show(@NonNull Message message, int duration) {
		// There is no message showed
		if (currentMessage == null) {
			int messageDelay = 0;
			if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
				//messageDelay = 750;
			}
			showInternal(message, duration, messageDelay);

		} else {
			// TODO: 03/01/2019 Решить, нужно ли разрешать дубликаты сообщений
			MessageRecord currentMessageRecord = currentMessage.getMessageRecord();
			MessageRecord newMessageRecord = message.getMessageRecord();

			if (newMessageRecord.equals(currentMessageRecord)) {
				Log.d(TAG, "show: trying to show the same message.. skipping");

			} else {
				if (!currentMessage.isDismissing()) {
					handler.removeCallbacksAndMessages(null);
					currentMessage.setOnDismissListener(() -> {
						currentMessage.setOnDismissListener(null);
						showInternal(message, duration, 0);
					});
					currentMessage.getCallback().dismiss();
				} else {
					Log.d(TAG, "show: we are in dismissing state");
				}
			}
		}
	}

	public void hide(@NonNull Message message) {
		message.getCallback().dismiss();
		message.showSystemUi();
	}

	private void showInternal(@NonNull Message message, int duration, int messageDelay) {
		currentMessage = checkNotNull(message);
		currentMessage.hideSystemUi();

		//		handler.postDelayed(() -> {
		//			Callback callback = currentMessage.getCallback();
		//			callback.show();
		//			if (duration > 0) {
		//				handler.removeCallbacksAndMessages(null);
		//				handler.postDelayed(() -> {
		//					//noinspection ConstantConditions
		//					if (callback != null) {
		//						callback.dismiss();
		//						currentMessage.showSystemUi();
		//						currentMessage = null;
		//					}
		//				}, duration);
		//			}
		//		}, messageDelay);

		Callback callback = currentMessage.getCallback();
		callback.show();
		if (duration > 0) {
			handler.removeCallbacksAndMessages(null);
			handler.postDelayed(() -> {
				//noinspection ConstantConditions
				if (callback != null) {
					callback.dismiss();
					currentMessage.showSystemUi();
					currentMessage = null;
				}
			}, duration);
		}
	}

	public void destroy() {
		handler.removeCallbacksAndMessages(null);
		currentMessage = null;
	}

	public interface Callback {

		void show();

		void dismiss();
	}
}

File: simple-message/src/main/java/ru/starksoft/simplemessage/SimpleMessageManager.kt
package ru.starksoft.simplemessage

import android.os.Handler
import android.os.Looper
import android.support.annotation.UiThread
import android.util.Log

internal class SimpleMessageManager private constructor() {

	private val handler = Handler(Looper.getMainLooper())
	private var currentMessage: Message? = null

	@UiThread
	fun show(message: Message, duration: Int) {
		if (currentMessage == null) {
			showInternal(message, duration, message.showMessageDelay())

		} else {
			// TODO: 03/01/2019 Решить, нужно ли разрешать дубликаты сообщений
			val currentMessageSafe = currentMessage!!
			val currentMessageRecord = currentMessageSafe.getMessageData()
			val newMessageRecord = message.getMessageData()

			if (newMessageRecord == currentMessageRecord) {
				Log.d(TAG, "show: trying to show the same message.. skipping")

			} else {
				if (!currentMessageSafe.isDismissing()) {
					handler.removeCallbacksAndMessages(null)
					currentMessageSafe.setOnDismissListener(object : Message.OnDismissListener {
						override fun onDismissed() {
							currentMessageSafe.setOnDismissListener(null)
							showInternal(message, duration, 0)
						}
					})
					currentMessageSafe.getCallback().dismiss()
				} else {
					Log.d(TAG, "show: we are in dismissing state")
				}
			}
		}
	}

	fun hide() {
		currentMessage?.apply {
			getCallback().dismiss()
			showSystemUi()
			currentMessage = null
		}
	}

	fun hide(message: Message) {
		message.getCallback().dismiss()
		message.showSystemUi()
	}

	private fun showInternal(message: Message, duration: Int, messageDelay: Long) {
		currentMessage = message
		val currentMessageSafe = currentMessage!!
		currentMessageSafe.hideSystemUi()

		handler.removeCallbacksAndMessages(null)
		handler.postDelayed({
								currentMessageSafe.getCallback().apply {
									show()
									if (duration > 0) {
										handler.removeCallbacksAndMessages(null)
										handler.postDelayed({
																dismiss()
																currentMessageSafe.showSystemUi()
																currentMessage = null

															}, duration.toLong())
									}
								}
							}, messageDelay)
	}

	fun destroy() {
		handler.removeCallbacksAndMessages(null)
		currentMessage = null
	}

	interface Callback {

		fun show()

		fun dismiss()
	}

	companion object {

		private const val TAG = "SimpleMessageManager"

		@JvmStatic
		val instance: SimpleMessageManager by lazy { SimpleMessageManager() }
	}
}
